### PR TITLE
New rule: postMessage-no-wildcard-targets

### PR DESCRIPTION
**Pull request by dmsnell**
Adds a new rule to warn when attempting to use the wildcard target
origin with `postMessage()`, which introduces a security vulnerability
allowing for data scraping by unintended parties.

For review I would especially appreciate some consideration for if I have used the right patterns here. Most of what I did was copy and paste work from other existing rules.

At first I wanted to limit this rule to invocations of `window.postMessage()` and `contentWindow.postMessage()` but then changed my mind because of the likelihood that someone might bind these and drop the parent object from the call, mixed with the unlikelihood that we would find the second parameter as the wildcard in these cases if they weren't actually calling `postMessage()` the way we expect.

**Note on failing build**
The tests were also failing on **master**

### CODE REVIEW

**Comment by aduth** [comment]
Tests failing on lint issues (whitespace consistency). There's one lint issue in an existing master file, but several more from this branch.

By convention, I think we should also stick to entirely lowercase rule names, either `postmessage-no-wildcard-targets` or `post-message-no-wildcard-targets`

Otherwise this is looking good. Thanks for tests, docs, and changelog notes!

**Comment by aduth** [comment]
Fixed lint issue from master in a75ac9f1113a8fd1c9b6e65b87f47847b3d419a9.

File: README.md [hunk 1]
```diff
@@ -52,6 +52,7 @@ Then configure the rules you want to use under the rules section.
 - [`jsx-classname-namespace`](docs/rules/jsx-classname-namespace.md): Ensure JSX className adheres to CSS namespace guidelines
 - [`jsx-gridicon-size`](docs/rules/jsx-gridicon-size.md): Enforce recommended Gridicon size attributes
 - [`import-docblock`](docs/rules/import-docblock.md): Enforce external, internal dependencies docblocks
+- [`post-message-no-wildcard-targets`](docs/rules/postMessage-no-wildcard-targets): Disallow using the wildcard '*' in `postMessage`
```
Comment (aduth) [inline README.md]: This is a broken link (needs `.md`)

Comment (dmsnell) [reply README.md]: added the `.me` and updated the name

File: tests/lib/rules/post-message-no-wildcard-targets.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+	config = { env: { es6: true } },
+	RuleTester = require( 'eslint' ).RuleTester;
+
+//------------------------------------------------------------------------------
+// Tests
+//------------------------------------------------------------------------------
+
+( new RuleTester( config ) ).run( 'post-message-no-wildcard-targets', rule, {
+	valid: [
+        { code: 'foo()' },
```
Comment (aduth) [inline tests/lib/rules/post-message-no-wildcard-targets.js]: Hmm, why aren't these failing on mixed spacing? Lint might not be covering tests.

File: tests/lib/rules/post-message-no-wildcard-targets.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        { code: 'foo( 1, \'*\' )' },
+        { code: 'postMessage( 1, \'test\' )' },
+        { code: 'postMessage( null, test )' },
+        { code: 'postMessage( true, test() )' }
+	],
+
+	invalid: [
```
Comment (aduth) [inline tests/lib/rules/post-message-no-wildcard-targets.js]: Seems fine after testing it myself, but might be worth adding a test for real-world example where postMessage is called on a window object, e.g. `frame.postMessage( 'bob', '*' )`

File: tests/lib/rules/post-message-no-wildcard-targets.js (hunk 2 above)
Comment (dmsnell) [reply tests/lib/rules/post-message-no-wildcard-targets.js]: oops. I didn't check after I used `elint --fix`

Comment (dmsnell) [reply tests/lib/rules/post-message-no-wildcard-targets.js]: replaced with tabs in 9554013

File: tests/lib/rules/post-message-no-wildcard-targets.js (hunk 3 above)
Comment (dmsnell) [reply tests/lib/rules/post-message-no-wildcard-targets.js]: more tests added in ef586eb

**Review by aduth — APPROVED** [review]
LGTM 👍

**Review by deBhal — CHANGES_REQUESTED** [review]
Looks good overall!

I was looking at that `node.arguments.length < 2` check, and wondering what the right way to handle it was, so I went looking for some examples where we were calling postMessage with a single value, and I found this gem for you, which would be nice to catch:

`target.postMessage(JSON.stringify(msg), o.origin || '*');`

For the single argument case, it's nearly always going to be a convenience function, and hopefully we'll catch it in the underlying convenience function itself, so I think ignoring it is the appropriate response. 👍

**Comment by deBhal** [comment]
Also, if you wanted to you could add `postMessage( '*', 'some.explicit.domain' )` to the "ok" list, just because that's the kind of implementation mistake that I might make.

**Comment by dmsnell** [comment]
> target.postMessage(JSON.stringify(msg), o.origin || '*');

ooh. I don't like that usage. actually I know there are ways around this but I'm trying to make a reasoned judgment on how far we should go

```js
const domain = '*';
target.postMessage( 'message', domain );
```

**Comment by dmsnell** [comment]
> Also, if you wanted to you could add postMessage( '*', 'some.explicit.domain' ) to the "ok" list, just because that's the kind of implementation mistake that I might make.

what do you mean by this @deBhal? just put it in the tests/examples to show that it's fine to send a message of `'*'`?

**Comment by deBhal** [comment]
Yes and yes.

I also dislike that `||` a lot and would complain about it gleefully, which might be coloring my judgement :)

You're totally right that a line needs to be drawn somewhere, and there's no point trying to catch someone who's maliciously trying to avoid the rule, but someone _might_ do it innocently, and it's not too hard to check programmatically (similar to the [binary concatenation we do in i18n-calypso]( https://github.com/Automattic/i18n-calypso/blob/d56d4929bdc2606862f256660febcf6599e44252/cli/preprocess-xgettextjs-match.js#L79 )). It's a pretty subjective call, though, and coming down on the "not checking" side seems justifiable if you want to do that.

( As an aside, I was idly wondering about a debug wrapper around postMessage to raise an error if '*' was ever used as the targetOrigin in `dev` through function/variable shenanigans, but that's really clearly out of scope for this PR. )

And yes, I was suggesting just adding the call to the `valid` list to document and verify that `'*'` is an acceptable message but not an acceptable `targetOrigin`, without making any other code changes.

**Comment by dmsnell** [comment]
> I also dislike that || a lot and would complain about it gleefully

for what it's worth I didn't see any occurrences of this inside Calypso.

I think adding a debug wrapper for `postMessage` isn't a terrible idea so we could catch runtime `*`s, but hopefully this linter will be enough of a motivation for developers to not use it.

**Pull request merged**